subsys/bluetooth/services/ots/ots_l2cap.c

```diff
@@ -102,11 +117,20 @@ static void l2cap_connected(struct bt_l2cap_chan *chan)
 
 static void l2cap_disconnected(struct bt_l2cap_chan *chan)
 {
+	struct bt_gatt_ots_l2cap *l2cap_ctx;
+
 	LOG_DBG("Channel %p disconnected", chan);
+
+	l2cap_ctx = CONTAINER_OF(chan, struct bt_gatt_ots_l2cap, ot_chan);
+
+	if (l2cap_ctx->closed) {
+		l2cap_ctx->closed(l2cap_ctx, chan->conn);
+	}
 }
 
 static const struct bt_l2cap_chan_ops l2cap_ops = {
 	.sent		= l2cap_sent,
+	.recv		= l2cap_recv,
```

Comment:
Is alloc_buf callback not needed for OTS? Is OTS using L2CAP CoC segmentation, or defining its own segmentation?

There is a configuration for BT_OTS_L2CAP_CHAN_RX_MTU, but that is not allowed to exceed the size of the L2CAP PDU MTU. So if OTS wants to received L2CAP SDUs segmented into multiple L2CAP PDUs then the callback should be defined and allocate buffers from a net buf pool sized from BT_OTS_L2CAP_CHAN_RX_MTU.

Reply:
I have to admit I am not very familiar with this aspect of Bluetooth. From what I see not having the `alloc_buf` callback is forcing the MTU to be truncated.
```
<wrn> bt_l2cap: Segmentation disabled but MTU > MPS, truncating MTU
```
It seems more correct to add the `alloc_buf` callback to allow the MTU to be dictated by the OTS layer.

Reply:
So you are sending and receiving L2CAP SDUs. The maximum size of the SDU is the MTU.
These packets are segmented into L2CAP PDUs. The maximum size of the PDU is the MPS.

By not having an alloc_buf callback then a single packet received in the l2cap_recv callback can me a maximum of MTU = MPS - 2.

By adding an alloc_buf callback then the application determines the maximum MTU by setting `chan->rx.mtu = CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU`.
The warning you saw there was the stack truncating the MTU to fit inside the MPS, so that option has not been  useful so far.

You could potentially skip having the alloc callback if the MPS is large enough for the OTS RX MTU.
`#if CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU <=  BT_L2CAP_SDU_RX_MTU`
 
The thing I couldn't quite work out by having a quick look at the OTS specification is whether OTS does segmentation on top of the L2CAP SDU again. But from what I'm seeing in this PR it looks like you can use offset and len to write an object that is larger than the SDU.
But from what I can tell there must be segmentation on top of the L2CAP SDU because the SDU length field is 16-bit, and the objects length field is 32-bit.

Reply:
Yes, for large objects (i.e. objects larger than the L2CAP MTU), OTS supports an additional level of fragmentation.

Reply:
I think for simplicity I will just leave the allocation callback.

Reply:
I have added conditional usage of the allocation callback only if `CONFIG_BT_OTS_L2CAP_CHAN_RX_MTU > BT_L2CAP_SDU_RX_MTU`.